### ISSUE TITLE
IntentModel with caching working much slower than without

### ISSUE THREAD

**Issue by jacekwasilewski** [RankSys/RankSys#10]
Hi @saulvargas,

I have found an interesting issue - if you use IntentModel with the caching option then e.g. xQuAD works waaaay longer than without caching - I assume with cache it should work faster, right?

Without caching - several seconds:

```
Jul 07, 2015 9:19:24 PM es.uam.eps.ir.ranksys.examples.RerankerExample lambda$main$4
INFO: START
Jul 07, 2015 9:19:48 PM es.uam.eps.ir.ranksys.examples.RerankerExample lambda$main$4
INFO: END
```

With caching - several minutes:

```
Jul 07, 2015 9:21:24 PM es.uam.eps.ir.ranksys.examples.RerankerExample lambda$main$4
INFO: START
Jul 07, 2015 9:34:12 PM es.uam.eps.ir.ranksys.examples.RerankerExample lambda$main$4
INFO: END
```

I had multithreading turned off, but when on it's still more or less like these results. And times are the times of reranking only.

Also I've check one more thing - I've counted how many times method `get()` of `IntentModel` class is called.
Without caching - the number of users, with caching much much more - I've stopped counting after a while but it was already several times more than the number of users.

Other models that inherit from `UserModel` are probably affected as well.

Jacek.

**Comment by saulvargas** [comment]
Hi @jacekwasilewski. These days I am a little busy. Hopefully, I will take a look at the issue during this week.

**Comment by jacekwasilewski** [comment]
No worries Saul - if I find a solution, I'll create a patch.

**Comment by saulvargas** [comment]
Hi @jacekwasilewski 

I took at look at the issue. First, a consideration: there is actually none (or little?) gain in caching when you are using a Model<U> just once. The caching feature was designed for cases when I had to use the same model over an over (for instance, testing xQuAD with a wide range of lambda values).

I did a quick check using MovieLens1M data and I found very little difference between caching and no caching in terms of execution time but, when considering memory usage...

WITHOUT CACHING
![caching-false](https://cloud.githubusercontent.com/assets/2616433/8598439/a6f6aec2-2652-11e5-8855-521963df1644.png)

WITH CACHING
![caching-true](https://cloud.githubusercontent.com/assets/2616433/8598450/ac7a7f68-2652-11e5-97fd-b00b6d8be82e.png)

I guess that the slowness is caused by the fact that you are running out of memory (because your dataset is presumably bigger or because you have thrown enough memory at the process) and therefore GC is used constantly. Could you please confirm?

I have also checked the number of times the IntentModel::get gets called. In principle, without caching this number should be equal to the number of users whose recommendations you are re-ranking, whereas in the other case it should be the number of users in your training data (or whatever set of users you use in the constructor of IntentModel). I checked that in two ways: with the profiler and with a manual count. Interestingly, the numbers in the first case are smaller than the second, being this last one equal to the number of training users, as expected. I'd blame some JVM optimisation for this (???). In any case, I suspect that what is causing the issue is the memory usage rather than this.

In conclusion, I think I will set caching to false in the examples, so that no one uses it as a default option.

**Comment by jacekwasilewski** [comment]
Hi @saulvargas,

I don't think the size of the data set was causing this problem - it was just MovieLens 1M and I have 32GB of memory.

I'm experiencing different behavior when it comes to the count of method calls - I'll try to dig into it. For me it looked more like "when I use Lazy (from UserModel constructor) to create a new model, I create all of them again". But I'm not sure about.

**Comment by jacekwasilewski** [comment]
Hi @saulvargas,

I think I have found an explanation - it is not how the `FeatureIntentModel` caches but how it is created and how xQuAD is created and then how it is used:

```
Map<String, Supplier<Reranker<Long, Long>>> rerankersMap = new HashMap<>();
rerankersMap.put("XQuAD", () -> {
    IntentModel<Long, Long, String> intentModel = new FeatureIntentModel<>(trainData, featureData);
    return new XQuAD<>(intentModel, lambda, cutoff, true);
});
rerankersMap.forEach((name, reranker) -> {
    System.out.println("Running " + name);
    String recOut = String.format("%s_%s", recIn, name);
    try (RecommendationFormat.Writer<Long, Long> writer = format.getWriter(recOut)) {
        format.getReader(recIn).readAll()
                .map(rec -> reranker.get().rerankRecommendation(rec, cutoff))
                .forEach(rerankedRecommendation -> {
                    try {
                        writer.write(rerankedRecommendation);
                    } catch (IOException ex) {
                        throw new UncheckedIOException(ex);
                    }
                });
    } catch (IOException e) {
        throw new UncheckedIOException(e);
    }
});
```

The problem really is here `reranker.get()` - our reranker is declared inside a Supplier function and it gets executed every time when we call `reranker.get()`, so for every user recommendation we create `IntentModel` and `XQuAD` objects.
If we use caching then in each computation of the Supplier function we collect all UserModels in the caching part - so if we count how many times `get(u)` gets called, we would get much more calls. And more memory would be allocated.

Probably I was not testing lambdas when I found that issue, so my `IntentModel` declaration was inside Supplier's function.

Jacek.

**Comment by saulvargas** [comment]
Great! Pull request accepted. Issue closed.
